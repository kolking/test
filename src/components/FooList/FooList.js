import React from 'react';
import { randomIndex, formatTime, formatDuration } from '../../helpers';

const FooList = ({ fooList, createFoo, updateFooCompleted }) => {
  const newFoo = () => {
    const speed = ['slow', 'fast', 'very fast'];
    const colour = ['blue', 'white', 'black', 'blue'];
    const size = ['tiny', 'small', 'medium', 'large', 'jumbo'];

    const fooItem = {
      speed: speed[randomIndex(speed.length)],
      colour: colour[randomIndex(colour.length)],
      size: size[randomIndex(size.length)],
      id: Math.random().toString(36).substring(7)
    }

    createFoo(fooItem.id, fooItem.colour, fooItem.size, fooItem.speed);
  }

  return (
    <section className="container">
      <hgroup className="container_header">
        <h1>List of foos</h1>
        <button type="button" onClick={newFoo}>Create New</button>
      </hgroup>
      <div className="container_content">
        {fooList && fooList.length > 0 ? (
          <ul className="foo-list">
            {fooList.map((item) => (
              <li key={item.id} className="foo-list_item">
                <time className="foo-list_item_time">
                  {formatTime(item.createdAt)}
                </time>
                <div className="foo-list_item_description">
                  {`[${item.id}] ${item.colour} - ${item.speed} - ${item.size}`}
                </div>
                {item.submittedAt ? (
                  <div className="foo-list_item_completed">
                    {`Completed in ${formatDuration(item.createdAt, item.submittedAt)}`}
                  </div>
                ) : (
                  <button type="button" onClick={() => updateFooCompleted(item.id)}>
                    Complete
                  </button>
                )}
                {item.errorMessage && (
                  <div className="error" data-message={item.errorMessage} />
                )}
              </li>
            ))}
          </ul>
        ) : (
          <p className="container_empty">There are no foos</p>
        )}
      </div>
    </section>
  )
}

export default FooList;