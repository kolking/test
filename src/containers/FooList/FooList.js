import { connect } from 'react-redux';
import { getAllFoos, createFoo, updateFooCompleted } from '../../reducers/foo/foo';
import FooList from '../../components/FooList';

const mapStateToProps = (state) => ({
  fooList: getAllFoos(state),
});

const mapDispatchToProps = {
  createFoo,
  updateFooCompleted,
};

export default connect(mapStateToProps, mapDispatchToProps)(FooList);