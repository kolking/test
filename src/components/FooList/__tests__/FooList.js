import React from 'react';
import TestRenderer from 'react-test-renderer';

/**
 * Under test
 */
import FooList from '../FooList';

const createElement = (props) => <FooList {...props} />;
const createRenderer = (props) => TestRenderer.create(createElement(props));

const fooList = [
  {
    id: 'uno',
    colour: 'red',
    size: 'big',
    speed: 'fast',
    createdAt: '2020-01-01T01:01:01Z',
  },
  {
    id: 'dos',
    colour: 'green',
    size: 'small',
    speed: 'slow',
    createdAt: '2020-02-02T02:02:02Z',
  },
  {
    id: 'tres',
    colour: 'blue',
    size: 'medium',
    speed: 'very fast',
    createdAt: '2020-03-03T03:03:03Z',
  },
];

it('should render empty list', () => {
  expect(createRenderer({})).toMatchSnapshot();
});

it('should render list of foos', () => {
  expect(createRenderer({ fooList })).toMatchSnapshot();
});

it('should call createFoo', () => {
  const createFoo = jest.fn();
  const tree = createRenderer({ createFoo });
  const button = tree.root.findAllByType('button')[0];

  button.props.onClick();

  expect(createFoo).toBeCalled();
});

it('should call toggleCompleteFoo', () => {
  const toggleCompleteFoo = jest.fn();
  const tree = createRenderer({ fooList, toggleCompleteFoo });
  const button = tree.root.findAllByType('button')[1];

  button.props.onClick();

  expect(toggleCompleteFoo).toBeCalled();
});
