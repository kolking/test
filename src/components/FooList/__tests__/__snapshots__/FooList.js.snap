// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render empty list 1`] = `
<section
  className="container"
>
  <hgroup
    className="container_header"
  >
    <h1>
      List of foos
    </h1>
    <button
      onClick={[Function]}
      type="button"
    >
      Create New
    </button>
  </hgroup>
  <div
    className="container_content"
  >
    <p
      className="container_empty"
    >
      There are no foos
    </p>
  </div>
</section>
`;

exports[`should render list of foos 1`] = `
<section
  className="container"
>
  <hgroup
    className="container_header"
  >
    <h1>
      List of foos
    </h1>
    <button
      onClick={[Function]}
      type="button"
    >
      Create New
    </button>
  </hgroup>
  <div
    className="container_content"
  >
    <ul
      className="foo-list"
    >
      <li
        className="foo-list_item"
      >
        <time
          className="foo-list_item_time"
        >
          5:01:01 AM
        </time>
        <div
          className="foo-list_item_description"
        >
          [uno] red - fast - big
        </div>
        <button
          onClick={[Function]}
          type="button"
        >
          Complete
        </button>
      </li>
      <li
        className="foo-list_item"
      >
        <time
          className="foo-list_item_time"
        >
          6:02:02 AM
        </time>
        <div
          className="foo-list_item_description"
        >
          [dos] green - slow - small
        </div>
        <button
          onClick={[Function]}
          type="button"
        >
          Complete
        </button>
      </li>
      <li
        className="foo-list_item"
      >
        <time
          className="foo-list_item_time"
        >
          7:03:03 AM
        </time>
        <div
          className="foo-list_item_description"
        >
          [tres] blue - very fast - medium
        </div>
        <button
          onClick={[Function]}
          type="button"
        >
          Complete
        </button>
      </li>
    </ul>
  </div>
</section>
`;
